nes-network/network/src/sender/control.rs

```diff
@@ -0,0 +1,453 @@
+
+async fn attempt_channel_registration<C: Communication + 'static>(
+    this_connection: ThisConnectionIdentifier,
+    channel: ChannelIdentifier,
+    channel_cancellation: CancellationToken,
+    queue: ChannelControlQueueListener,
+    channel_tx: ChannelEstablishRequest,
+    controller: NetworkingConnectionController,
+    communication: C,
+) {
+    let retry = ExponentialBackoff::from_millis(2)
+        .max_delay_millis(500)
+        .map(jitter);
+
+    let target_channel_identifier = Retry::spawn(
+        retry,
+        async || -> core::result::Result<ConnectionIdentifier, RetryError<Error>> {
+            let (tx, rx) = oneshot::channel();
+            if channel_tx.send((channel.clone(), tx)).await.is_err() {
+                return Err(RetryError::Permanent("NetworkService Shutdown".into()));
+            }
+
+            let Ok(result) = rx.await else {
+                return Err(RetryError::Permanent("NetworkService Shutdown".into()));
+            };
+
+            match result {
+                EstablishChannelResult::Ok(channel_connection_identifier) => {
+                    Ok(channel_connection_identifier)
+                }
+                EstablishChannelResult::ChannelReject => Err(Transient {
+                    err: "Channel was rejected".into(),
+                    retry_after: None,
+                }),
+                EstablishChannelResult::BadConnection(_, _) => Err(Transient {
+                    err: "Bad connection".into(),
+                    retry_after: None,
+                }),
+            }
+        },
+    )
+        .await;
+
+    let Ok(target_channel_identifier) = target_channel_identifier else {
+        return;
+    };
+
+    create_channel_handler(
+        this_connection,
+        target_channel_identifier,
+        channel,
+        channel_cancellation,
+        queue,
+        communication,
+        controller,
+    );
+}
+
+async fn connection_handler<C: Communication + 'static>(
+    this_connection: ThisConnectionIdentifier,
+    target_connection: ConnectionIdentifier,
+    controller: NetworkingConnectionController,
+    listener: NetworkingConnectionControlListener,
+    communication: C,
+) -> Result<()> {
+    let (request_connection, mut await_connection_request) = tokio::sync::mpsc::channel::<
+        oneshot::Sender<(
+            ControlChannelSenderReader<C::Reader>,
+            ControlChannelSenderWriter<C::Writer>,
+        )>,
+    >(1);
+
+    let (channel_registration_request_handler, mut await_channel_registration_request) =
+        tokio::sync::mpsc::channel::<(ChannelIdentifier, oneshot::Sender<EstablishChannelResult>)>(
+            10,
+        );
+
+    // Connection Keepalive Task:
+    // This task is responsible for establishing connections to the target_connection.
+    // It implements exponential backoff.
+    // Once a connection has been established and transfers the Reader/Writer pair to `connection_tx`
```

Comment:
Outdated comment, there is no `connection_tx`.

Reply:
I updated it